lib/tests/streamlit/state/session_state_test.py

```diff
@@ -692,3 +646,58 @@ def test_setattr_reserved_key(self, _):
     def test_delattr_reserved_key(self, _):
         with pytest.raises(StreamlitAPIException):
             delattr(self.lazy_session_state, self.reserved_key)
+
+
+@given(state=stst.session_state())
+def test_compact_idempotent(state):
+    assert state._compact() == state._compact()._compact()
+
+
+@given(state=stst.session_state())
+def test_compact_len(state):
+    assert len(state) >= len(state._compact())
+
+
+@given(state=stst.session_state())
+def test_compact_presence(state):
+    assert state.items() == state._compact().items()
+
+
+@given(m=stst.session_state())
+def test_mapping_laws(m):
+    assert len(m) == len(m.keys()) == len(m.values()) == len(m.items())
+    assert [value for value in m.values()] == [m[key] for key in m.keys()]
+    assert [item for item in m.items()] == [(key, m[key]) for key in m.keys()]
+
+
+@given(
+    m=stst.session_state(),
+    key=stst.user_key,
+    value1=hst.integers(),
+    value2=hst.integers(),
+)
+def test_map_set_set(m, key, value1, value2):
+    m[key] = value1
+    l1 = len(m)
+    m[key] = value2
+    assert m[key] == value2
+    assert len(m) == l1
+
+
+@given(m=stst.session_state(), key=stst.user_key, value1=hst.integers())
+def test_map_set_del(m, key, value1):
+    m[key] = value1
+    l1 = len(m)
+    keys = m.keys() - {key}
+    del m[key]
+    assert key not in m
+    assert len(m) < l1
+    # for k in keys:
+    #     assert k in m
```

Comment:
Could we be more exact here and `assert len(m) == l1 - 1`?

Also, is there something strange breaking with the for loop so that it's currently commented out?

Reply:
The for loop actually doesn't hold for session state as it currently functions, because for keyed widgets we have to represent it in `.keys()` as either the key or the widget id, but then deleting via the other one will cause it to no longer be present in the dict